picolibc: Update to version 1.8.7

keith-packard opened the pull request:
Integrate new upstream release

The associated sdk-ng work is happening on https://github.com/zephyrproject-rtos/sdk-ng/tree/topic-picolibc-1.8.7

keith-packard committed:
picolibc: Update to version 1.8.7

Integrate new upstream release.
Include strnlen_s and aarch64 cmake fixes.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
posix: Avoid stepping on picolibc clockid_t and timer_t

picolibc sets _CLOCKID_T_DECLARED and _TIMER_T_DECLARED, not
__clockid_t_defined and __timer_t_defined.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
posix: Avoid redefining TZNAME_MAX

When _POSIX_C_SOURCE is defined by a source file, a POSIX-conforming C
library will define TZNAME_MAX in limits.h. Avoid redefining this
in Zephyr's posix_features.h.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
posix: Don't require struct itimerspec in posix/time.h

struct itimerspec is only required to be available when building with
_POSIX_C_SOURCE defined. Applications including posix/time.h without
defining _POSIX_C_SOURCE will need to have a forward declaration of struct
itimerspec even if they don't use the functions requiring that type.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/common: Define _POSIX_C_SOURCE to gain access to POSIX functions

ctime.c and localtime_r_utc.c may use POSIX-only functions; to ensure those
are visible from a POSIX-conforming C library, define _POSIX_C_SOURCE in
these source files.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
posix: Define _POSIX_C_SOURCE to gain access to POSIX functions

options/fs.c and options/timer.c both use POSIX-only functions. To ensure
those symbols are visible from the underlying C library, define
_POSIX_C_SOURCE.

Signed-off-by: Keith Packard <keithp@keithp.com>

cfriedt commented on lib/libc/common/source/time/ctime.c:
```diff
@@ -4,6 +4,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#undef _POSIX_C_SOURCE
```
I understand why the kludge is necessary. I was hoping to address this particular issue preemptively almost 9 months ago :'(

https://github.com/zephyrproject-rtos/zephyr/pull/67132

cfriedt replied:
I'm happy to approve as-is, but @keith-packard, if you feel like we could use the Kconfig, I would be more than happy to pursue that PR again.

keith-packard replied:
Up to you; we can revert this change if that happens.

keith-packard committed:
posix: Need sys/types.h to get suseconds_t defined

Picolibc's sys/_timeval.h (which Zephyr shouldn't be using) needs
sys/types.h to be included first so that __suseconds_t is defined.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
posix: options/shm.c needs _POSIX_C_SOURCE

This file uses PATH_MAX, which is defined in limits.h when _POSIX_C_SOURCE
is set.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
net: dns: Ignore GCC warnings in mdns_responder.c

GCC complains about struct sockaddr accesses due to the various
address-family-specific variants being of differing sizes. Let's not
mess with code (which looks correct), just silence the compiler.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
subsys/shell: cmd_date_set uses gmtime_r, a POSIX function

Define _POSIX_C_SOURCE for this file to make sure gmtime_r is visible.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests, samples: Define _POSIX_C_SOURCE where needed

Various tests and samples use POSIX-only functions. Define _POSIX_C_SOURCE
to ensure they are visible.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/c_lib: Don't test undefined behavior of asctime and asctime_r

These functions requre non-null parameters; passing them NULL
values is undefined. Remove tests which are not valid.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/c_lib: Ensure TZ value is UTC for ctime/ctime_r tests

When testing these functions with CONFIG_NATIVE_LIBC, force the TZ
environment variable to UTC to ensure they produce the expected results.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
subsys/llext: Allow .exported_sym to be between other data sections

There's no reason for the .exported_sym data to always land at one end of a
data region; the order of sections depends on the whim of the compiler and
assembler.

Ignore overlaps between that region and other data regions.

Signed-off-by: Keith Packard <keithp@keithp.com>

jukkar commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -35,6 +35,18 @@ LOG_MODULE_REGISTER(net_mdns_responder, CONFIG_MDNS_RESPONDER_LOG_LEVEL);
 #include "net_private.h"
 
+/*
+ * GCC complains about struct sockaddr accesses due to the various
+ * address-family-specific variants being of differing sizes. Let's not
+ * mess with code (which looks correct), just silence the compiler.
```
Interesting. I have not seen these issues so far, what gcc version is this?

Also could you add `net: ` prefix to the dns commit subject, it is just easier to find it later if needed.
So `net: dns: ....`

keith-packard replied:
Looks like I'm using gcc 14.2.0 on Debian unstable. I'lve fixed the dns: commit message, now to go fix more build failures.

keith-packard committed:
posix: sysconf.c uses POSIX values

We need to add _POSIX_C_SOURCE to this file as it uses POSIX values from
limits.h including IOV_MAX, CHILD_MAX and ARG_MAX.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
Given the number of changes in the POSIX code, I'd love to get @cfriedt to give this another glance before it gets merged?

cfriedt commented on lib/libc/common/Kconfig:
```diff
@@ -8,26 +8,30 @@ config COMMON_LIBC_ABORT
 
 config COMMON_LIBC_ASCTIME_R
 	bool "Thread-safe version of asctime()"
 	default y if POSIX_THREAD_SAFE_FUNCTIONS
```
These lines should be removed, and instead in a POSIX Kconfig, we should use a pattern like:

```
config POSIX_C_LANG_SUPPORT_R
    select LIBC_COMMON_ASCTIME_R if ! TC_PROVIDES_POSIX_C_LANG_SUPPORT_R
    ...
```

That way picolibc could just select `TC_PROVIDES_POSIX_C_LANG_SUPPORT_R` and cover all of those functions at once. That might be closer in intent to the way that the TC_PROVIDES macros gate files being built in cmakelists.txt and we don't need to use `depends on ! PICOLIBC`

keith-packard replied:
Had to remove a couple of redundant 'default y if POSIX_THREAD_SAFE_FUNCTIONS' clauses. Looks like a cleaner plan, thanks!

keith-packard committed:
libc/common: Remove redundant Kconfig POSIX_THREAD_SAFE_FUNCTIONS uses

POSIX_THREAD_SAFE_FUNCTIONS already selects POSIX_C_LANG_SUPPORT_R and
POSIX_C_LANG_SUPPORT_R alread selects all of the COMMON_LIBC_*_R values.
That makes the 'default y if POSIX_THREAD_SAFE_FUNCTIONS' clauses in these
symbols redundant.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
lib/posix: Only enable Zephyr _r APIs when toolchain doesn't

Check TC_PROVIDES_POSIX_C_LANG_SUPPORT_R before selecting
the COMMON_LIBC_*_R APIs.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/picolibc: Don't replace picolibc time _r functions

Select TC_PROVIDES_POSIX_C_LANG_SUPPORT_R to keep Zephyr from
including the common libc implementation of the various _r APIs.

Signed-off-by: Keith Packard <keithp@keithp.com>

cfriedt reviewed (approved):
LGTM.

I'll respin #67132 as well, so that we can have a way to avoid peppering `_POSIX_C_SOURCE` throughout the code, for consistency.
